hs/test/UnitTests.hs

```diff
@@ -301,46 +403,51 @@ testSpendNotOwnedUTxO =
               (Set.fromList [TxIn genesisId 1])
               [ TxOut aliceAddr (Coin 232)]
               []
+              Map.empty
               (Coin 768)
               (Slot 100)
-    aliceWit = makeWitness alicePay txbody
+    aliceWit = makeWitness  txbody alicePay
     tx = TxWits txbody (Set.fromList [aliceWit])
-  in ledgerState [tx] @?= Left [InsufficientWitnesses]
+  in ledgerState [tx] @?= Left [MissingWitnesses, UnneededWitnesses]
 
-testInvalidTransaction :: Assertion
-testInvalidTransaction =
+testWitnessWrongUTxO :: Assertion
+testWitnessWrongUTxO =
   let
     txbody = Tx
               (Set.fromList [TxIn genesisId 1])
               [ TxOut aliceAddr (Coin 230)]
               []
+              Map.empty
               (Coin 770)
               (Slot 100)
     tx2body = Tx
-              (Set.fromList [TxIn genesisId 0])
-              [ TxOut aliceAddr (Coin 19230)]
+              (Set.fromList [TxIn genesisId 1])
+              [ TxOut aliceAddr (Coin 230)]
               []
+              Map.empty
               (Coin 770)
-              (Slot 100)
-    aliceWit = makeWitness alicePay tx2body
+              (Slot 101)
+    aliceWit = makeWitness  tx2body alicePay
     tx = TxWits txbody (Set.fromList [aliceWit])
-  in ledgerState [tx] @?= Left [InsufficientWitnesses]
+  in ledgerState [tx] @?= Left [ InvalidWitness
+                               , MissingWitnesses
+                               , UnneededWitnesses ]
 
 testEmptyInputSet :: Assertion
 testEmptyInputSet =
```

Comment:
Now that we have refunds, I can make this test more in the original spirit, since now we can fail just precisely on this one error.